internal/ssa2ast/func.go

```diff
@@ -0,0 +1,1161 @@
+			}
+			stmt = defineVar(i, sliceExpr)
+		case *ssa.SliceToArrayPointer:
+			castExpr, err := fc.tc.Convert(i.Type())
+			if err != nil {
+				return err
+			}
+			xExpr, err := fc.convertSsaValue(i.X)
+			if err != nil {
+				return err
+			}
+			stmt = defineVar(i, ah.CallExpr(&ast.ParenExpr{X: castExpr}, xExpr))
+		case *ssa.Store:
+			addrExpr, err := fc.convertSsaValue(i.Addr)
+			if err != nil {
+				return err
+			}
+			valExpr, err := fc.convertSsaValue(i.Val)
+			if err != nil {
+				return err
+			}
+			stmt = ah.AssignStmt(&ast.StarExpr{X: addrExpr}, valExpr)
+		case *ssa.TypeAssert:
+			valExpr, err := fc.convertSsaValue(i.X)
+			if err != nil {
+				return err
+			}
+
+			assertTypeExpr, err := fc.tc.Convert(i.AssertedType)
+			if err != nil {
+				return err
+			}
+
+			if i.CommaOk {
+				valName, valType, valHasRefs := fc.tupleVarNameAndType(i, 0)
+				okName, okType, okHasRefs := fc.tupleVarNameAndType(i, 1)
+				if valHasRefs {
+					astFunc.Vars[valName] = valType
+				}
+				if okHasRefs {
+					astFunc.Vars[okName] = okType
+				}
+
+				stmt = &ast.AssignStmt{
+					Lhs: []ast.Expr{ast.NewIdent(valName), ast.NewIdent(okName)},
+					Tok: token.ASSIGN,
+					Rhs: []ast.Expr{&ast.TypeAssertExpr{X: valExpr, Type: assertTypeExpr}},
+				}
+			} else {
+				stmt = defineVar(i, &ast.TypeAssertExpr{X: valExpr, Type: assertTypeExpr})
+			}
+		case *ssa.UnOp:
+			valExpr, err := fc.convertSsaValue(i.X)
+			if err != nil {
+				return err
+			}
+			if i.CommaOk {
+				if i.Op != token.ARROW {
+					return fmt.Errorf("unary operator %s in %v: %w", i.Op, instr, ErrUnsupported)
+				}
+
+				valName, valType, valHasRefs := fc.tupleVarNameAndType(i, 0)
+				okName, okType, okHasRefs := fc.tupleVarNameAndType(i, 1)
+				if valHasRefs {
+					astFunc.Vars[valName] = valType
+				}
+				if okHasRefs {
+					astFunc.Vars[okName] = okType
+				}
+
+				stmt = &ast.AssignStmt{
+					Lhs: []ast.Expr{ast.NewIdent(valName), ast.NewIdent(okName)},
+					Tok: token.ASSIGN,
+					Rhs: []ast.Expr{&ast.UnaryExpr{
+						Op: token.ARROW,
+						X:  valExpr,
+					}},
+				}
+			} else if i.Op == token.MUL {
+				stmt = defineVar(i, &ast.StarExpr{X: valExpr})
+			} else {
+				stmt = defineVar(i, &ast.UnaryExpr{Op: i.Op, X: valExpr})
+			}
+		case *ssa.MakeClosure:
+			anonFunc := i.Fn.(*ssa.Function)
+			anonFuncName, err := fc.getAnonFunctionName(anonFunc)
+			if err != nil {
+				return err
+			}
+			if anonFuncName == nil {
+				return fmt.Errorf("make closure for non anon func %s: %w", anonFunc.Name(), ErrUnsupported)
+			}
+
+			callExpr := &ast.CallExpr{Fun: anonFuncName}
+			for _, freeVar := range i.Bindings {
+				varExr, err := fc.convertSsaValue(freeVar)
+				if err != nil {
+					return err
+				}
+				callExpr.Args = append(callExpr.Args, varExr)
+			}
+
+			stmt = defineVar(i, callExpr)
+		case *ssa.RunDefers, *ssa.DebugRef:
+			// ignored
+			continue
+		default:
+			return fmt.Errorf("instruction %v: %w", instr, ErrUnsupported)
+		}
```

Comment:
Is this something which happens under normal operation or should this panic?

Reply:
https://github.com/burrowers/garble/pull/752#discussion_r1234420220